staging/operator-lifecycle-manager/pkg/controller/operators/packageserver_controller.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package operators
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/operators/manifests"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/utils/pointer"
+
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	name                = "packageserver"
+	infrastructureName  = "cluster"
+	defaultReplicaCount = int32(2)
+	singleReplicaCount  = int32(1)
+)
+
+// PackageServerReconciler reconciles the PackageServer deployment object
+type PackageServerReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+
+	Namespace string
+	// TODO(tflannag): Should this be a custom struct instead?
+	// TODO(tflannag): Add a mutex here?
+	// TODO(tflannag): Add an event recorder?
+	HighlyAvailableMode bool
+}
+
+//+kubebuilder:rbac:groups=config.openshift.io,resources=infrastructures,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=config.openshift.io,resources=infrastructures/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=config.openshift.io,resources=infrastructures/finalizers,verbs=update
```

Comment:
nit: we can probably dial these back a notch -- and add some of the core types this needs to play with too.